robotoff/prediction/category/neural/keras_category_classifier_3_0/preprocessing.py

```diff
@@ -0,0 +1,272 @@
+        ingredient_taxonomy, add_synonym_combinations=True
+    )
+
+
+def generate_inputs_from_product(product: JSONType, ocr_texts: list[str]) -> JSONType:
+    """Generate inputs for v3 category predictor model.
+
+    :param product: the product dict, the `product_name` and `ingredients`
+        fields are used, if provided
+    :param ocr_texts: a list of detected OCR texts, one per image
+    :return: a dict containing inputs for v3 category predictor model
+    """
+    ingredient_taxonomy = get_ingredient_taxonomy()
+    ingredient_processor = get_ingredient_processor()
+
+    inputs = {
+        "product_name": product.get("product_name", ""),
+        "ingredients_tags": transform_ingredients_input(
+            product.get("ingredients", []), ingredient_taxonomy
+        )
+        or [""],
+        "ingredients_ocr_tags": extract_ocr_ingredients(
+            ocr_texts, processor=ingredient_processor, debug=False
+        )
+        or [""],
+    }
+
+    nutriments = product.get("nutriments", {})
+    for nutriment_name in NUTRIMENT_NAMES:
+        inputs[nutriment_name] = transform_nutrition_input(
+            nutriments.get(f"{nutriment_name.replace('_', '-')}_100g"),
+            nutriment_name=nutriment_name,
+        )
+    return inputs
+
+
+def remove_untaxonomized_values(value_tags: list[str], taxonomy: Taxonomy) -> list[str]:
+    return [value_tag for value_tag in value_tags if value_tag in taxonomy]
+
+
+def transform_ingredients_input(
+    ingredients: list[dict], taxonomy: Taxonomy
+) -> list[str]:
+    # Only keep nodes of depth=1 (i.e. don't keep sub-ingredients)
+    # While sub-ingredients may be interesting for classification, enough signal is already
+    # should already be present in the main ingredient, and it makes it more difficult to
+    # take ingredient order into account (as we don't know if sub-ingredient #2 of
+    # ingredient #1 is more present than sub-ingredient #1 of ingredient #2)
+    return remove_untaxonomized_values(
+        [get_tag(ingredient["id"]) for ingredient in ingredients], taxonomy
+    )
+
+
+def transform_nutrition_input(value: Optional[float], nutriment_name: str) -> float:
+    """Transform nutritional values before model inference.
+
+    This function returns:
+
+    - -1, if the value is missing
+    - -2, if the value is obviously or suspiciously wrong (<0 or >=101 for all
+        nutriments except energy, <0 or >=3800 kcal for energy)
+    - the input value otherwise
+
+    :param value: the float value
+    :param nutriment_name: the name of the nutriment
+    :return: the transformed nutriment value
+    """
+    if value is None:
+        return -1
+
+    if nutriment_name == "energy-kcal":
+        if value >= 3800 or value < 0:
+            # Too high to be true
+            return -2
+
+    elif value < 0 or value >= 101:
+        # Remove invalid values
+        return -2
+
+    return value
+
+
+MULTIPLE_SPACES_REGEX = re.compile(r" {2,}")
```

Comment:
Indeed, overall it would be better, even if here it doesn't make any difference as Google Cloud OCR only returns standard whitespace in OCR results.
Concerning every non-bug comment (except better documentation) to preprocessing.py, I'm going to close them, as I don't want to add training/inference skew by introducing without noticing subtle changes in the preprocessing code.